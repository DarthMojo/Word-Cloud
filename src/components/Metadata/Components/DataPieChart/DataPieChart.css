.PieChart {
    display: flex;
    justify-content: center;
    flex-direction: column;
    align-items:center;
}

.ChartTitle{
    font-size: 1em;
    display: flex;
    justify-content: center;
    align-self: center;
    margin-bottom: 5%;
    text-decoration: underline;
}