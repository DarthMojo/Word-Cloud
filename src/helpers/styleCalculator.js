/**
 * 1. Each topic should have one of 6 different text sizes,
 * with the most popular topics largest, and least popular smallest
 *
 * 2.A topic with a sentiment score > 60 should be displayed in green
 *
 * 3.A topic with a sentiment score < 40 should be displayed in red
 *
 * 4.Other topics should be displayed in grey
 */
import type Dimension from "../config/types";
import { sentimentScoreMin,
    sentimentScoreMax,
    defaultColor,
    size } from "../config/static";


/**
 *
 * @type {{getTopicSize: (function(*, *)),
 * getTopicColor: (function(*)),
 * windowSize: (function(*=)),
 * rotation: (function(): number)}}
 */
const styleCalculator = {
    /**
     *
     * @param wordPopularity
     * @param popularityRange
     * @return {number}
     */
    getTopicSize: (wordPopularity, popularityRange): number => {
        const unitRange = ((popularityRange.max - popularityRange.min) / size.max) || 0;
        const sizeValue: number = Array.from({ length: size.max }, (v, k): number => k + 1)
            .find((sizeIndex: number): number =>
                wordPopularity <= popularityRange.min + (sizeIndex * unitRange));
        return Math.round((sizeValue / size.max) * 100);
    },
    /**
     *
     * @param sentimentScore
     * @return String
     */
    getTopicColor: (sentimentScore): string => {
        if (sentimentScore > sentimentScoreMax.score) {
            return sentimentScoreMax.color;
        } else if (sentimentScore < sentimentScoreMin.score) {
            return sentimentScoreMin.color;
        }
        return defaultColor;
    },
    /**
     *
     * @param factor
     * @return {{width: number, height: number}}
     */
    windowSize: (factor = 1): Dimension => {
        return {
            width: Math.round(window.innerWidth * factor),
            height: Math.round(window.innerHeight * factor)
        };
    },
    /**
     *@return {number}
     */
    rotation: (): number => (~~(Math.random() * size.max) - size.max * 0.5) * 30
};
export default styleCalculator;
