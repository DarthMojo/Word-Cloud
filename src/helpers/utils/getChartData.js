/**
 *
 * @param values
 * @return {Array}
 */
const getChartData = (values) => {
    const data = [];
    for (const [key, value] of Object.entries(values)) {
        data.push({ x: key, y: value });
    }
    return data;
};
export default getChartData;
