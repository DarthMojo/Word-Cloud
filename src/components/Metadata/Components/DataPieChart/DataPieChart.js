import React from "react";
import type { Element } from "react";
import { PieChart } from "react-easy-chart";
import { chartSizeFactor, chartMargin } from "../../../../config/static";
import "./DataPieChart.css";

type DataPieChartType = {
    data: Array<Object>;
    dimensions: Object;
    chartTitle: string
};

/**
 *
 * @param dimensions
 * @param data
 * @param chartTitle
 * @return {Element}
 * @constructor
 */
const DataPieChart = ({ dimensions, data, chartTitle }: DataPieChartType): Element<any> => {
    const processedData = data.map((value, index) =>
        ({ key: value.x, value: value.y }));
    return (<div className={"PieChart"}>
        <div className={"ChartTitle"}>{chartTitle}</div>
        <PieChart
            size={(dimensions.height * chartSizeFactor)}
            innerHoleSize={(dimensions.height * chartSizeFactor) / 2}
            data={processedData}
            labels
            margin={chartMargin}
        />
    </div>);
};

export default DataPieChart;
