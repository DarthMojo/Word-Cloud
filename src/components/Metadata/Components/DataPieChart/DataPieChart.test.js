import React from "react";
import DataPieChart from "./DataPieChart";
import getChartData from "../../../../helpers/utils/getChartData";
import data from "../../../../config/data/topics.json";
import { shallowToJson } from "enzyme-to-json";
import styleCalculator from "../../../../helpers/styleCalculator";

test("render DataPieChart()", () => {
    const dimensions = styleCalculator.windowSize();

    const wordData = data.topics.find((wordItem): Object<*> =>
        wordItem.id === "1751295897__Berlin");

    const wrapper = shallow(
        <DataPieChart
            data={getChartData(wordData.sentiment)}
            dimensions={dimensions}/>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
});
