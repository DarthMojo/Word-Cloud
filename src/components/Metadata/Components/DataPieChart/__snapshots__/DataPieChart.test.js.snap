// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render DataPieChart() 1`] = `
<div
  className="PieChart"
>
  <div
    className="ChartTitle"
  />
  <PieChart
    clickHandler={[Function]}
    data={
      Array [
        Object {
          "key": "negative",
          "value": 3,
        },
        Object {
          "key": "neutral",
          "value": 133,
        },
        Object {
          "key": "positive",
          "value": 29,
        },
      ]
    }
    innerHoleSize={288}
    labels={true}
    margin={
      Object {
        "bottom": 50,
        "left": 50,
        "right": 50,
        "top": 50,
      }
    }
    mouseMoveHandler={[Function]}
    mouseOutHandler={[Function]}
    mouseOverHandler={[Function]}
    padding={2}
    size={576}
    styles={Object {}}
  />
</div>
`;
